src/runtime/runtime.ts

```diff
@@ -221,12 +253,81 @@ export class Runtime {
     // we could eliminate it if the Manifest object takes care of this.
     const id = `in-memory-${Math.floor((Math.random()+1)*1e6)}.manifest`;
     // TODO(sjmiles): this is a virtual manifest, the fileName is invented
-    const localOptions = {id, fileName: `./${id}`, loader};
-    return Manifest.parse(content, {...localOptions, ...options});
+    const opts = {id, fileName: `./${id}`, loader, ...options};
+    return Manifest.parse(content, opts);
+  }
+
+  async parseFile(path: string, options?): Promise<Manifest> {
+    const content = await this.loader.loadResource(path);
+    const opts = {id: path, fileName: path, loader: this.loader, ...options};
+    return this.parse(content, opts);
+  }
+
+  async resolveRecipe(arc: Arc, recipe: Recipe): Promise<Recipe | null> {
+    if (this.normalize(recipe)) {
+      if (recipe.isResolved()) {
+        return recipe;
+      }
+      const resolver = new RecipeResolver(arc);
+      const plan = await resolver.resolve(recipe);
+      if (plan && plan.isResolved()) {
+        return plan;
+      }
+      warn('failed to resolve:\n', (plan || recipe).toString({showUnresolved: true}));
+    }
+    return null;
   }
 
+  normalize(recipe: Recipe): boolean {
+    if (Runtime.isNormalized(recipe)) {
+      return true;
+    }
+    const errors = new Map();
+    if (recipe.normalize({errors})) {
+      return true;
+    }
+    warn('failed to normalize:\n', errors, recipe.toString());
+    return false;
+  }
+
+  static isNormalized(recipe: Recipe): boolean {
+    return Object.isFrozen(recipe);
+  }
+
+  // TODO(sjmiles): redundant vs. newArc, but has some impedance mismatch
+  // strategy is to merge first, unify second
+  async spawnArc({id, serialization, context, composer, storage, portFactories}: SpawnArgs): Promise<Arc> {
+    const params = {
+      id: IdGenerator.newSession().newArcId(id),
+      fileName: './serialized.manifest',
+      serialization,
+      context,
+      storageKey: storage || 'volatile',
+      slotComposer: composer,
+      pecFactories: [this.pecFactory, ...(portFactories || [])],
+      loader: this.loader,
+      // TODO(sjmiles): maybe doesn't belong here, but empirically it's
+      // wanted in (almost?) all Arc instantiations
+      inspectorFactory: devtoolsArcInspectorFactory
+    };
+    return serialization ? Arc.deserialize(params) : new Arc(params);
+  }
+
+  // static interface for the default runtime environment
+
   static async parse(content: string, options?): Promise<Manifest> {
     return this.getRuntime().parse(content, options);
   }
 
+  static async parseFile(path: string, options?): Promise<Manifest> {
```

Comment:
I think it would be helpful for these two functions to have different type signatures or very different names.

I'd have to look this up most times I see it to check which is which.

Maybe `parseFromPath`?

Reply:
Hrm, not clear why `parseFromPath` is superior to `parseFile`, but I really don't mind changing it.

Also, I believe this function is not invoked from anywhere in reality, because it's possible to call `parse` on content like `import 'foo.arcs'`, which is akin to `parseFile('foo.arcs')` but the former can take advantage of path-mapping goodies. 

I almost deleted it completely, but there is a subtle difference between the results of the two methods (`parseFile` puts objects at the top level, `parse` will only have objects starting at `imports[0]`).